courses/api.py

```diff
@@ -776,48 +777,59 @@ def generate_course_run_certificates():
         return
 
     for run in course_runs:
-        edx_grade_user_iter = exception_logging_generator(
-            get_edx_grades_with_users(run)
-        )
-        created_grades_count, updated_grades_count, generated_certificates_count = (
-            0,
-            0,
-            0,
+        generate_course_run_certificates_for_course(run)
+
+
+def generate_course_run_certificates_for_course(run):
+    """
+    For given course run syncs grades and
+    if certificate_available_date is in the past, also generates certificates.
+
+    Args:
+        run (models.CourseRun): a course run instance.
+    """
+    edx_grade_user_iter = exception_logging_generator(get_edx_grades_with_users(run))
+    created_grades_count, updated_grades_count, generated_certificates_count = (
+        0,
+        0,
+        0,
+    )
+    for edx_grade, user in edx_grade_user_iter:
+        course_run_grade, created, updated = ensure_course_run_grade(
+            user=user, course_run=run, edx_grade=edx_grade, should_update=True
         )
-        for edx_grade, user in edx_grade_user_iter:
-            course_run_grade, created, updated = ensure_course_run_grade(
-                user=user, course_run=run, edx_grade=edx_grade, should_update=True
+
+        if created:
+            created_grades_count += 1
+        elif updated:
+            updated_grades_count += 1
+
+        # Check certificate generation eligibility
+        #   1. For self_paced course runs we generate certificates right away irrespective
+        #   of certificate_available_date
+        #   2. For others course runs we generate the certificates if the certificate_available_date of course run
+        #   has passed
+        now = now_in_utc()
+        if run.is_self_paced or (
+            run.certificate_available_date and run.certificate_available_date <= now
```

Comment:
I was writing some [documentation](https://docs.google.com/document/d/1kw-ChTQK2YZZr8QjexhQX7P-IH5wwQeXNr4xxG0i2r8/edit#heading=h.j7mv17u378xp) last week about how certificates are generated and it took me way too long to understand all the different clauses we have scattered around regarding `certificate_available_date`.  Would it make sense to remove this `if` statement completely, and add the check of `is_self_paced` within `get_certificate_grade_eligible_runs`?

Right now we check certificate_available_date in the following places:
handle_update_course_run: certificate_available_date exists and is in the past.
generate_course_run_certificates_for_course: certificate_available_date exists and is in the past, or the run is self paced.
get_certificate_grade_eligible_runs: certificate_available_date does not exist or is in the past (default 24 hours).

I've tried to list out all of the scenarios in which course run certificates should be created in the documentation I linked.  I think there is a lot of overlap being handled by each one of these functions that we could consolidate.

Reply:
Yes, I think it could be refactored.